### DISCUSSION TITLE
Time-dependent arc cost

### DISCUSSION THREAD

**Discussion by FirstPotatoOnMars** [google/or-tools#4230] [Routing (and legacy CP) questions]
Is it possible to realize the time-dependent  arc cost in ortools VRP solver? 
For example, the travel time of arc AB is 5 min when it is 8:00, while the travel time changes to 10 min at 9:30.

**Comment by lperron** [comment]
Asked and answered multiple times. 

Short answer: you cannot.

**Reply by FirstPotatoOnMars** [reply]
😥sad

**Reply by lperron** [reply]
There are work around. The slightly longer answer is that data precision makes that computation worthless, and that solving in multiple phases is enough.

**Comment by 0nyr** [comment]
An update for future readers, since this discussion still comes up when searching for time-dependent travel times in OR-Tools: the situation is unchanged in 2026, and the cumul-dependent transit API that appeared in recent releases is non-functional and slated for removal (see #5070, where the team confirms "This code does not work"). If your problem is genuinely time-dependent (travel time a function of departure time, FIFO), there is now an open-source option for the research-grade core of that class: [Kayros](https://github.com/0nyr/kayros) (MIT, `pip install kayros`), exact and anytime for duration-minimization TDVRP(TW) over piecewise-linear travel-time functions, with certified optima on the public [MAMUT-routing](https://mamut-routing.onyr.net/) benchmarks. Honest scoping: it is a C++/Python research solver with a duration objective, not a drop-in OR-Tools replacement.
